irc.go

```diff
@@ -54,6 +54,89 @@ func (ms *modeSet) Apply(s string) error {
 	return nil
 }
 
+type channelModeType byte
+
+const (
+	modeTypeA channelModeType = iota
+	modeTypeB
+	modeTypeC
+	modeTypeD
+)
+
+var stdChannelModes = map[byte]channelModeType{
+	'b': modeTypeA,
+	'e': modeTypeA,
+	'I': modeTypeA,
+	'k': modeTypeB,
+	'l': modeTypeC,
+	'i': modeTypeD,
+	'm': modeTypeD,
+	'n': modeTypeD,
+	's': modeTypeD,
+	't': modeTypeD,
+}
+
+func ApplyChannelModes(modeTypes map[byte]channelModeType, modes map[byte]string, modeStr string, arguments ...string) error {
+	nextArgument := 0
+	var plusMinus byte
+	for i := 0; i < len(modeStr); i++ {
+		mode := modeStr[i]
+		if mode == '+' || mode == '-' {
+			plusMinus = mode
+			continue
+		}
+		if plusMinus != '+' && plusMinus != '-' {
+			return fmt.Errorf("malformed modestring %q: missing plus/minus", modeStr)
+		}
+
+		mt, ok := modeTypes[mode]
+		if !ok {
+			continue
+		}
+		if mt == modeTypeB || (mt == modeTypeC && plusMinus == '+') {
```

Comment:
This can probably be simplified to something among those lines:

```go
if plusMinus == '+' {
	var param string
	switch mt {
	case modeTypeB, modeTypeC:
		// some sentitive arguments […]
		if nextArgument < len(arguments) {
			…
		}
	case modeTypeC, modeTypeD:
		// Mode has no param
	}
	modes[mode] = param
} else {
	delete(modes, mode)
}
```

Reply:
Not sure I can simply further. See `nextArgument++`.